drivers/gpio/gpio_stmpe1600.c

```diff
@@ -0,0 +1,297 @@
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master = drvdata->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	int ret = 0;
+	uint16_t transfer_data = sys_cpu_to_le16(value);
+
+	LOG_DBG("STMPE1600[0x%02X]: write REG[0x%02X..0x%02X] = %04x",
+		i2c_addr, reg, reg + 1, value);
+
+	ret = i2c_burst_write(
+		i2c_master, i2c_addr, reg,
+		(uint8_t *)&transfer_data, sizeof(transfer_data)
+		);
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: write error REG[0x%02X..0x%02X]: %d",
+			i2c_addr, reg, reg + 1, ret);
+	}
+	return ret;
+}
+
+static int read_reg16(const struct device *dev, uint8_t reg, uint16_t *value)
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master = drvdata->i2c_master;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	int ret = 0;
+	uint16_t transfer_data = 0;
+
+	LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X]", i2c_addr, reg, reg + 1);
+
+	ret = i2c_burst_read(
+		i2c_master, i2c_addr, reg,
+		(uint8_t *) &transfer_data, sizeof(transfer_data)
+		);
+
+	if (ret != 0) {
+		LOG_ERR("STMPE1600[0x%02X]: read error REG[0x%02X..0x%02X]: %d",
+			i2c_addr, reg, reg + 1, ret);
+	} else {
+		*value = sys_le16_to_cpu(transfer_data);
+		LOG_DBG("STMPE1600[0x%02X]: read REG[0x%02X..0x%02X] => %04x",
+			i2c_addr, reg, reg + 1, *value);
+	}
+	return ret;
+}
+
+static int set_pin_dir(const struct device *dev, gpio_pin_t pin, gpio_flags_t flags)
+{
+	uint16_t reg_dir = 0, reg_state = 0;
+	int ret = 0;
+
+	ret = read_reg16(dev, REG_GPDR_LSB, &reg_dir);
+	if (ret) {
+		return ret;
+	}
+	ret = read_reg16(dev, REG_GPSR_LSB, &reg_state);
+	if (ret) {
+		return ret;
+	}
+
+	if ((flags & GPIO_INPUT) != 0U) {
+		reg_dir &= ~BIT(pin);
+	} else {
+		reg_dir |= BIT(pin);
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			reg_state |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			reg_state &= ~BIT(pin);
+		}
+	}
+
+	ret = write_reg16(dev, REG_GPDR_LSB, reg_dir);
+	if (ret) {
+		return ret;
+	}
+	ret = write_reg16(dev, REG_GPSR_LSB, reg_state);
+	return ret;
+}
+
+static int stmpe1600_init(const struct device *dev)
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master;
+	uint16_t chip_id = 0;
+	int ret = 0;
+
+	LOG_DBG("STMPE1600[0x%02X] init", config->i2c_slave_addr);
+
+	i2c_master = device_get_binding((char *)config->i2c_master_dev_name);
+	if (!i2c_master) {
+		return -EINVAL;
+	}
+	drvdata->i2c_master = i2c_master;
+
+	k_sem_init(&drvdata->lock, 1, 1);
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+	ret = read_reg16(dev, REG_CHIP_ID_LSB, &chip_id);
+	k_sem_give(&drvdata->lock);
```

Comment:
Driver initialization happens during boot time, in a single thread. There is no need to take and release the semaphore.